### PR TITLE
Fixes #27032 - improve logging and timeout

### PR DESCRIPTION
**Pull request by lzap**
Foreman proxy DNS module performs record resolve operating during create request, but there is not a single log line about it. It is worth adding also timing information because this can be cruical for finding why some DNS requests are slow.

### CODE REVIEW

File: modules/dns_common/dns_common.rb [hunk 1]
```diff
@@ -17,9 +19,28 @@ def initialize(server = nil, ttl = nil)
     end
 
     def resolver
+      logger.warn("Proxy::Dns::Record#resolver has been deprecated and will be removed, use get_dns_resource(s)")
```
Comment (ekohl) [inline modules/dns_common/dns_common.rb]: :-1: to removing this. It's a good place to abstract it out. smart_proxy_dns_route53 also uses this to ignore the DNS servers since Route53 doesn't really provide stable IPs to query. That's why it uses the system resolver. You'd break this functionality.

https://github.com/theforeman/smart_proxy_dns_route53/blob/86e6393d75192e991f2f2128c57ad94471d8180d/lib/smart_proxy_dns_route53/dns_route53_main.rb#L52-L54

File: modules/dns_common/dns_common.rb [hunk 2]
```diff
@@ -17,9 +19,28 @@ def initialize(server = nil, ttl = nil)
...
       Resolv::DNS.new(:nameserver => @server)
     end
 
+    def get_dns_resource(name, typeclass)
+      result = nil
+      duration = time_spent_in_ms do
+        result = Resolv::DNS.new(:nameserver => @server).getresource(name, typeclass)
```
Comment (ekohl) [inline modules/dns_common/dns_common.rb]:
```suggestion
        result = resolver.getresource(name, typeclass)
```

File: modules/dns_common/dns_common.rb [hunk 3]
```diff
@@ -17,9 +19,28 @@ def initialize(server = nil, ttl = nil)
...
+      end
+      logger.info "Resolved #{name} via system resolver in #{duration.round(2)} ms"
```
Comment (ekohl) [inline modules/dns_common/dns_common.rb]: This is debug info
```suggestion
      logger.debug "Resolved #{name} via system resolver in #{duration.round(2)} ms"
```

File: modules/dns_common/dns_common.rb [hunk 4]
```diff
@@ -17,9 +19,28 @@ def initialize(server = nil, ttl = nil)
...
+      result
+    end
+
+    def get_dns_resources(name, typeclass)
+      result = nil
+      duration = time_spent_in_ms do
+        result = Resolv::DNS.new(:nameserver => @server).getresources(name, typeclass)
```
Comment (ekohl) [inline modules/dns_common/dns_common.rb]:
```suggestion
        result = resolver.getresources(name, typeclass)
```

File: modules/dns_common/dns_common.rb [hunk 5]
```diff
@@ -17,9 +19,28 @@ def initialize(server = nil, ttl = nil)
...
+      end
+      logger.info "Resolved #{name} via system resolver in #{duration.round(2)} ms"
```
Comment (ekohl) [inline modules/dns_common/dns_common.rb]:
```suggestion
      logger.debug "Resolved #{name} via system resolver in #{duration.round(2)} ms"
```

**Review by ekohl — CHANGES_REQUESTED** [review]

File: modules/dns_common/dns_common.rb (hunk 2 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: This method is deprecated and files a warning into the logger.

File: modules/dns_common/dns_common.rb (hunk 3 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: This is intentional, I actually want to be very verbose here.

File: modules/dns_common/dns_common.rb (hunk 4 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: Ditto.

File: modules/dns_common/dns_common.rb (hunk 5 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: Ditto.

File: modules/dns_common/dns_common.rb (hunk 1 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: No, I am not breaking the compatibility, I am deprecating the functionality with a plan to remove (break) it. That's a difference.

One possible solution to avoid modification of plugins is to provide a decorator object that would keep the API. If you insist, I can do that, however I think that it's hacky solution. No other plugin uses that and I am willing to patch route53 until the method is deprecated.

**Comment by lzap** [comment]
Thanks for review, however what I am doing here is all intentional. I've spent tremeneous time debugging this issue with two customers affected by this, it's not yet clear if this was the real cause for DNS requests to be served with 90 second delays, but this will make much easier.

I also intentionally want to inform users via an INFO message that there is a resolution being done via *system* resolver. I think this is worth an INFO message, every request already does several INFO messages and this is only for the DNS creation ones.

File: modules/dns_common/dns_common.rb (hunk 1 above)
Comment (ekohl) [reply modules/dns_common/dns_common.rb]: IMHO this method still makes sense. I'd be ok with deprecating it and making it private if there's a good reason solution to route53, but duplicating the creation of a resolver object is something I'd like to avoid.

File: modules/dns_common/dns_common.rb (hunk 3 above)
Comment (ekohl) [reply modules/dns_common/dns_common.rb]: It's actually also incorrect. It uses specific nameservers from the configuration so it's not using the system resolver.

I'm against such verbose timing logging. Didn't we introduce a timing measuring API to store this kind of data in an easily parsable way?

**Comment by ekohl** [comment]
> there is a resolution being done via _system_ resolver

But it's not the system resolver. The idea is that you can specify the IP of your authoritative nameserver and it would query that. If you're running your host on 192.168.100.2 and an infoblox server on 192.168.100.3 then you should specify that nameserver IP. If you're using the system resolver, it's a misconfiguration.

This still isn't perfect and for API-based systems it should really query those (PowerDNS, Route53) to detect whether there are conflicts and the whole system needs an overhaul, but that's out of scope now.

File: modules/dns_common/dns_common.rb (hunk 1 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: But what do you recommend then? I believe that defining a better contract (two methods) is so much better than just a method which creates new instance.

Comment (ekohl) [reply modules/dns_common/dns_common.rb]: Use all 3 methods. The new methods should be recommended. I'd also copy the caching of the instance that route53 does so it becomes a method that lazy loads the resolver instance.

File: modules/dns_common/dns_common.rb (hunk 3 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb]: You are right that this does not use system resolver, there is a fallback to localhost tho. I will correct the string, also I am willing to change this to debug if you think it's really not something that should appear every DNS creation.

I am not aware of any kind of timing API, if you mean telemetry API that has been added only to Foreman core. Of course, once we add it here (and it's on my long-term todo) this will be included.

**Comment by lzap** [comment]
Yeah, I figured out that this is not system resolver, gonna correct that and this also explains it. Also the customer confirmed that this is the root cause of their issue.

Looks like route 53 does not even use the helper, so I am actually changing my mind and I will provide a decorator around Resolv Ruby object with measuring and a helper that plugins can optionally use.

https://github.com/theforeman/smart_proxy_dns_route53/blob/master/lib/smart_proxy_dns_route53/dns_route53_main.rb

**Comment by lzap** [comment]
I am looking in our installer codebase and I believe that the "server" DNS module setting is not parametrized, so effectively this is localhost only. Can you confirm?

**Comment by lzap** [comment]
How about this, completely rewritten from scratch. There is a Resolv decorator available now, nothing is changing for plugins and it only creates one debug line, however if the duration is longer it will create a warning with more details for easier troubleshooting. Unfortuateny there is no way to pull nameservers out of the Resolv object other than inspect or reading private attribute.

This is how log would look like when there was a slow query:

```
2019-06-13T10:50:00 49db72b5 [W] Slow DNS query (getresources) in 12030.66 ms: ["test228.ibx.lan", Resolv::DNS::Resource::IN::A]
2019-06-13T10:50:00 49db72b5 [D] Resolver used: #<Resolv::DNS:0x000055555759fc20 @mutex=#<Thread::Mutex:0x000055555759fbd0>, @config=#<Resolv::DNS::Config:0x000055555759fb30 @mutex=#<Thread::Mutex:0x000055555759fae0>, @config_info={:nameserver=>"192.168.201.2"}, @initialized=true, @timeouts=[10], @nameserver_port=[["192.168.201.2", 53]], @search=[[#<Resolv::DNS::Label::Str home>, #<Resolv::DNS::Label::Str lan>]], @ndots=1>, @initialized=true>
```

File: modules/dns_common/dns_common.rb [hunk 6]
```diff
@@ -8,16 +9,20 @@ class Collision < RuntimeError; end
...
 
     attr_reader :server, :ttl
 
-    def initialize(server = nil, ttl = nil)
+    def initialize(server = nil, ttl = nil, resolv_timeout = nil)
       @server = server || "localhost"
-      @ttl    = ttl || "86400"
+      @ttl = ttl || "86400"
+      @resolv_timeout = resolv_timeout || 10
     end
```
Comment (lzap) [inline modules/dns_common/dns_common.rb:20]: Optionally timeout can be now specified, Ruby default is quite tricky to understand.

File: modules/dns_common/dns_common.rb [hunk 7]
```diff
@@ -8,16 +9,20 @@ class Collision < RuntimeError; end
...
 
-    def resolver
-      Resolv::DNS.new(:nameserver => @server)
+    def resolver(override_nameserver = @server)
+      resolv = Resolv::DNS.new(:nameserver => override_nameserver)
+      resolv.timeouts = @resolv_timeout
+      ResolvLogDecorator.new(resolv)
     end
```
Comment (lzap) [inline modules/dns_common/dns_common.rb:24]: Also nameserver can be overriden by plugins now, I could add this to the initializer above but I think here it's also fine since this is actually "constructor" of the resolver object.

**Comment by lzap** [comment]
Fixed rubocop and tests, this is now ready for re-review.

**Review by ekohl — COMMENTED** [review]
I still think this is just making it hard to follow. The core problem is that this class is overloaded and has too many concerns. That was already the case before this change.

I do like the wrapper that extracts the resolving, but maybe it should go further. I would actually like to introduce a `ConflictChecker` interface. Using plain DNS is one implementation, but networked services like Route53 and Infoblox should be able to introduce their own that may not rely on DNS queries at all.

**Comment by lzap** [comment]
Honestly, I can't commit for refactoring of this class with breaking changes, however I like the idea of plugins being easily able to provide own checking interface. How about changing the code in more simple way introducing method `do_get` alongside `do_create` and `do_remove` with the default implementation using Ruby DNS resolver. This would give ability to override this and also would not break existing plugins at all.

The class is not that bad, it's quite long because of all the record types, but if you strip it to the bare minimum code, it's basically CREATE and DELETE with lots of deprecated stuff. Maybe if we remove deprecated stuff that would help to clean it a bit too.

**Comment by ekohl** [comment]
So I started with https://github.com/theforeman/smart-proxy/pull/659.

And you're right: my goal was to just have a very simple interface which passes just the right DNS data and let implementations handle the specifics. https://github.com/theforeman/smart-proxy/pull/507 as an early attempt, but I soon realized we need a V2 API that doesn't have the oddness that comes from the REST API. The creation of A and PTR records is very confusing because it deals with IPs rather than just DNS names and their values. I haven't picked this up because I can't really commit to it either (though the whole capabilities framework was the first step in optionally providing it).

I'm going to need a bit to think about how to do this cleanly but I think there's a cleaner way. Will try to get back to you tomorrow.

**Comment by lzap** [comment]
Here is a rough idea (untested):

```diff
diff --git a/modules/dns_common/dns_common.rb b/modules/dns_common/dns_common.rb
index 31c1c69..e9318e5 100644
--- a/modules/dns_common/dns_common.rb
+++ b/modules/dns_common/dns_common.rb
@@ -19,12 +19,37 @@ module Proxy::Dns
       @resolv_timeout = resolv_timeout || 10
     end
 
+    # Methods which MUST be implemented by providers
+    def do_create(name, value, type)
+      raise(Proxy::Dns::Error, "Creation of #{type} not implemented")
+    end
+
+    def do_remove(name, type)
+      raise(Proxy::Dns::Error, "Deletion of #{type} not implemented")
+    end
+
+    # Methods which CAN be implemented by providers
+    def do_get_resources(name, type)
+      # must return an array
+      do_get_via_ruby_resolver(name, type)
+    end
+
+    # The rest is common code for all providers
     def resolver(override_nameserver = @server)
       resolv = Resolv::DNS.new(:nameserver => override_nameserver)
       resolv.timeouts = @resolv_timeout
       ResolvLogDecorator.new(resolv)
     end
 
+    def do_get_via_ruby_resolver(name, type)
+      resolver.getresources(name, type)
+    end
+
+    def do_get_resource(name, type)
+      do_get_resources.each(name, type) {|resource| return resource}
+      raise Resolv::ResolvError.new("DNS result has no information for #{name}")
+    end
+
     def create_srv_record(service, value)
       do_create(service, value, 'SRV')
     end
@@ -73,10 +98,6 @@ module Proxy::Dns
       end
     end
 
-    def do_create(name, value, type)
-      raise(Proxy::Dns::Error, "Creation of #{type} not implemented")
-    end
-
     def remove_a_record(fqdn)
       do_remove(fqdn, "A")
     end
@@ -97,10 +118,6 @@ module Proxy::Dns
       do_remove(name, "PTR")
     end
 
-    def do_remove(name, type)
-      raise(Proxy::Dns::Error, "Deletion of #{type} not implemented")
-    end
-
     def dns_find(key)
       logger.warn(%q{Proxy::Dns::Record#dns_find has been deprecated and will be removed in future versions of Smart-Proxy.
                       Please use ::Proxy::Dns::Record#get_name or ::Proxy::Dns::Record#get_address instead.})
@@ -138,13 +155,13 @@ module Proxy::Dns
     end
 
     def get_resource_as_string(value, resource_type, attr)
-      resolver.getresource(value, resource_type).send(attr).to_s
+      do_get_resource(value, resource_type).send(attr).to_s
     rescue Resolv::ResolvError
       false
     end
 
     def get_resource_as_string!(value, resource_type, attr)
-      resolver.getresource(value, resource_type).send(attr).to_s
+      do_get_resource(value, resource_type).send(attr).to_s
     rescue Resolv::ResolvError
       raise Proxy::Dns::NotFound.new("Cannot find DNS entry for #{value}")
     end
@@ -195,14 +212,14 @@ module Proxy::Dns
         raise Proxy::Dns::Error.new("Not an IP Address: '#{ip}'")
       end
 
-      resources = resolver.getresources(fqdn, type)
+      resources = do_get_resources(fqdn, type)
       return -1 if resources.empty?
       return 0 if resources.any? {|r| IPAddr.new(r.address.to_s) == ip_addr }
       1
     end
 
     def record_conflicts_name(fqdn, type, content)
-      resources = resolver.getresources(fqdn, type)
+      resources = do_get_resources(fqdn, type)
       return -1 if resources.empty?
       return 0 if resources.any? {|r| r.name.to_s.casecmp(content) == 0 }
       1
```

**Comment by ekohl** [comment]
Have a look at my deprecation PR. Most of these functions are unused.

Perhaps we can have a more generic `record_conflicts(name, type, content)`. Then the infoblox plugin can just query the HTTP API to see if there's a conflict.

Alternatively, if the creation can be done in a single REST call it should override the `create_*_record` methods to perform it as part of that.

**Comment by lzap** [comment]
I am completely fine with a separate commit deprecating stuff, but I will likely need to backport this change if we agree on an interface/method for Infoblox to provide conflict resulution on its own so I'd appreciate if this commit could go first.

**Comment by lzap** [comment]
> Perhaps we can have a more generic record_conflicts(name, type, content). Then the infoblox plugin can just query the HTTP API to see if there's a conflict.

Isn't better to keep conflict handling code in the common module and define the contract as minimal possible? Just small operations, CRUD basically.

**Comment by lzap** [comment]
I see that dropping of some unused code has started already, so now I don't mind testing and merging your other deprecation patch.

**Comment by lzap** [comment]
@ekohl I have a suggestion, how about I turn this PR into generic resolver API for all core and plugins. On every place where we do `Resolv.new` we'd use this new decorator which provides additional logging. Then we can also work on improving the DNS API towards better support for DNS conflict detection.

**Comment by ekohl** [comment]
That was roughly what I was thinking about yes.

**Comment by lzap** [comment]
Rebased and pushed a WIP version, one question let me find the code.

File: lib/proxy/helpers.rb [hunk 8]
```diff
@@ -64,13 +64,24 @@ def parse_json_body
...
   # reverse lookup an IP address while verifying it via forward resolv
   def remote_fqdn(forward_verify=true)
     ip = request.env['REMOTE_ADDR']
     log_halt 403, 'could not get remote address from environment' if ip.empty?
 
     begin
-      dns = Resolv.new
+      # TODO - shall we actually make this dns_resolv?
+      dns = resolv
       fqdn = dns.getname(ip)
```
Comment (lzap) [inline lib/proxy/helpers.rb:88]: Here in this remote_fqdn method I realized that we don't actually do DNS resolve but generic resolve, this will use any OS provider (kerberos, NSS, DNS). Do we actually want to keep this behavior or shell I change this to strictly DNS only?

File: modules/dns_common/dns_common.rb (hunk 7 above)
Comment (lzap) [reply modules/dns_common/dns_common.rb:24]: I still want to keep this, I might need to do a hotfix/quickfix for affected customer because this whole DNS refactoring and legacy code removal might block me from doing full Infoblox conflict query rewrite backport.

File: lib/proxy/helpers.rb (hunk 8 above)
Comment (lzap) [reply lib/proxy/helpers.rb:88]: Ping @ekohl

Comment (ekohl) [reply lib/proxy/helpers.rb:88]: Sorry, I missed this. I'm leaning to the current resolv. I think the value of using NSS is actually pretty big and something we should be preferring unless we really need DNS.

**Comment by lzap** [comment]
Then I amended and removed the note in the code.

**Comment by lzap** [comment]
I have just found that the reverse lookup in bootServer method does not work properly, because we throw away exception it was actually not working. Gonna fix this and add tests.

File: modules/dhcp_common/isc/omapi_provider.rb [hunk 9]
```diff
@@ -195,22 +195,25 @@ def poap_options_statements(options)
...
           ns = "\\\"#{server}\\\""
         end
       end
       "next-server = #{ns};"
     end
 
-    def ip2hex ip
-      ip.split(".").map{|i| "%02x" % i }.join(":")
+    def ip2hex(ip)
+      ip.to_s.split(".").map{|i| "%02x" % i }.join(":")
```
Comment (lzap) [inline modules/dhcp_common/isc/omapi_provider.rb]: This was actually not working, because we swallowed the exception we haven't noticed. Resolver returns IPv4 object which has no `split` method. Probably the reason why we added reverse search into Foreman core?! Not sure. It now has tests and this PR also adds nice logging around this so it's very obvious what is going on.

**Comment by lzap** [comment]
Rebased, fixed issue with reverse DNS search and added couple of tests around it.

File: modules/dhcp_common/isc/omapi_provider.rb (hunk 9 above)
Comment (ekohl) [reply modules/dhcp_common/isc/omapi_provider.rb]: Good find. Almost feels like it deserves a separate Redmine issue.

File: lib/proxy/logging_resolv.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    private
+
+    def call_with_timing(method, *args)
+      result = nil
+      duration = time_spent_in_ms do
+        result = @resolv.send(method, *args)
+      end
+      if duration < 9000
```
Comment (ekohl) [inline lib/proxy/logging_resolv.rb]: Small nit: I like constants for this. Perhaps a `SLOW_DNS_QUERY_MS` definition? Note I'm not asking for a config value, just a constant in the class.

File: modules/dhcp_common/isc/omapi_provider.rb [hunk 11]
```diff
@@ -195,22 +195,25 @@ def poap_options_statements(options)
...
-        ns = ip2hex validate_ip(server)
+        ns = ip2hex(validate_ip(server))
       rescue
         begin
-          ns = ip2hex Resolv.new.getaddress(server)
-        rescue
-          logger.warn "Failed to resolve IP address for #{server}"
+          logger.info "Next-server option not IPv4, trying to resolve PTR record for #{server}"
+          ns = ip2hex(dns_resolv.getaddress(server))
+        rescue Exception => e
```
Comment (ekohl) [inline modules/dhcp_common/isc/omapi_provider.rb]: Is there a `ResolvException` or similar we can catch just as a generic resolution failed? That could use `logger.warn` and this could log at the exception level.

**Review by ekohl — COMMENTED** [review]
Overall it's looking good. I do feel like the bootserver fixes could use a separate Redmine issue for visibility in the changelog.

File: lib/proxy/logging_resolv.rb (hunk 10 above)
Comment (lzap) [reply lib/proxy/logging_resolv.rb]: Sure, it's probably an overkill but with all the issues, cases and confused users around DNS queries I want to be as clear as possible when this happens.

File: modules/dhcp_common/isc/omapi_provider.rb (hunk 11 above)
Comment (lzap) [reply modules/dhcp_common/isc/omapi_provider.rb]: Yes it's `Resolv::ResolvError`.

**Comment by lzap** [comment]
Rebased.

File: modules/dhcp_common/isc/omapi_provider.rb [hunk 12]
```diff
@@ -195,22 +195,25 @@ def poap_options_statements(options)
...
+    def bootServer(server)
       begin
-        ns = ip2hex validate_ip(server)
+        ns = ip2hex(validate_ip(server))
       rescue
         begin
-          ns = ip2hex Resolv.new.getaddress(server)
-        rescue
-          logger.warn "Failed to resolve IP address for #{server}"
+          logger.info "Next-server option not IPv4, trying to resolve PTR record for #{server}"
```
Comment (ekohl) [inline modules/dhcp_common/isc/omapi_provider.rb]: PTR is reverse DNS but here it's forward DNS.

**Comment by lzap** [comment]
Force-pushed, fixed the logging message.

**Review by ekohl — CHANGES_REQUESTED** [review]
Please extract the omapi fixes to a separate PR. It's incomplete (you missed `ztp_options_statements` which will fail if `nextServer` is a DNS name) and holding the logging PR.

**Comment by lzap** [comment]
Can you give it another review? I removed the omapi bits and moved them to: https://github.com/theforeman/smart-proxy/pull/682

**Review by ekohl — APPROVED** [review]
Thanks!
